README.md
# prefect-earthdata

<p align="center">
    <!--- Insert a cover image here -->
    <!--- <br> -->
    <a href="https://pypi.python.org/pypi/prefect-earthdata/" alt="PyPI version">
        <img alt="PyPI" src="https://img.shields.io/pypi/v/prefect-earthdata?color=0052FF&labelColor=090422"></a>
    <a href="https://github.com/giorgiobasile/prefect-earthdata/" alt="Stars">
        <img src="https://img.shields.io/github/stars/giorgiobasile/prefect-earthdata?color=0052FF&labelColor=090422" /></a>
    <a href="https://pypistats.org/packages/prefect-earthdata/" alt="Downloads">
        <img src="https://img.shields.io/pypi/dm/prefect-earthdata?color=0052FF&labelColor=090422" /></a>
    <a href="https://github.com/giorgiobasile/prefect-earthdata/pulse" alt="Activity">
        <img src="https://img.shields.io/github/commit-activity/m/giorgiobasile/prefect-earthdata?color=0052FF&labelColor=090422" /></a>
    <br>
    <a href="https://prefect-community.slack.com" alt="Slack">
        <img src="https://img.shields.io/badge/slack-join_community-red.svg?color=0052FF&labelColor=090422&logo=slack" /></a>
    <a href="https://discourse.prefect.io/" alt="Discourse">
        <img src="https://img.shields.io/badge/discourse-browse_forum-red.svg?color=0052FF&labelColor=090422&logo=discourse" /></a>
</p>

Visit the full docs [here](https://giorgiobasile.github.io/prefect-earthdata) to see additional examples and the API reference.

Prefect integrations with NASA Earthdata, taking advantage of the [`earthaccess`](https://nsidc.github.io/earthaccess/) library.


## Getting started

`prefect-earthdata` provides a Prefect credentials block and a few tasks to interact with
NASA Earthdata. It does so by leveraging the `earthaccess` library and its API.

Loading the `EarthdataCredentials` block with your credentials, corresponds to calling the [`earthaccess.login()`](https://nsidc.github.io/earthaccess/user-reference/api/api/#earthaccess.api.login) function.

After that, all other `earthaccess` functions can be directly used, without having to login again.

Nevertheless, a few tasks are provided to help taking full advantage of Prefect's observability features.

### Search and download on NASA Earthdata

```python
from prefect import flow, get_run_logger

from prefect_earthdata.credentials import EarthdataCredentials
from prefect_earthdata.tasks import download, search_data


@flow(log_prints=True)
def example_earthdata_download_flow():

    logger = get_run_logger()

    earthdata_credentials = EarthdataCredentials.load("earthdata-credentials")

    granules = search_data(
        earthdata_credentials,
        count=1,
        short_name="ATL08",
        bounding_box=(-92.86, 16.26, -91.58, 16.97),
    )

    logger.info(f"File URL: {granules[0].data_links()[0]}")

    download_path = "/tmp"

    logger.info(f"Downloading data to {download_path}")
    files = download(
        credentials=earthdata_credentials,
        granules=granules,
        local_path=download_path,
    )
    logger.info(f"Downloaded files: {files}")

    return granules, files

example_earthdata_download_flow()
```

Output:

```python
21:07:16.603 | INFO    | prefect.engine - Created flow run 'cheerful-peacock' for flow 'example-earthdata-download-flow'
21:07:18.121 | INFO    | Flow run 'cheerful-peacock' - Created task run 'search_data-0' for task 'search_data'
21:07:18.121 | INFO    | Flow run 'cheerful-peacock' - Executing 'search_data-0' immediately...
21:07:19.642 | INFO    | Task run 'search_data-0' - You're now authenticated with NASA Earthdata Login
21:07:19.644 | INFO    | Task run 'search_data-0' - Using token with expiration date: 09/01/2023
21:07:20.243 | INFO    | Task run 'search_data-0' - Using environment variables for EDL
21:07:21.432 | INFO    | Task run 'search_data-0' - Granules found: 760
21:07:22.211 | INFO    | Task run 'search_data-0' - Finished in state Completed()
21:07:22.225 | INFO    | Flow run 'cheerful-peacock' - File URL: https://data.nsidc.earthdatacloud.nasa.gov/nsidc-cumulus-prod-protected/ATLAS/ATL08/005/2018/11/05/ATL08_20181105083647_05760107_005_01.h5
21:07:22.226 | INFO    | Flow run 'cheerful-peacock' - Downloading data to /tmp
21:07:22.586 | INFO    | Flow run 'cheerful-peacock' - Created task run 'download-0' for task 'download'
21:07:22.587 | INFO    | Flow run 'cheerful-peacock' - Executing 'download-0' immediately...
21:07:24.529 | INFO    | Task run 'download-0' - We are already authenticated with NASA EDL
21:07:28.677 | INFO    | Task run 'download-0' -  Getting 1 granules, approx download size: 0.0 GB
QUEUEING TASKS | : 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 2016.49it/s]
PROCESSING TASKS | : 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:11<00:00, 11.45s/it]
COLLECTING RESULTS | : 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 25115.59it/s]
21:07:40.388 | INFO    | Task run 'download-0' - Finished in state Completed()
21:07:40.392 | INFO    | Flow run 'cheerful-peacock' - Downloaded files: ['ATL08_20181105083647_05760107_005_01.h5']
21:07:41.243 | INFO    | Flow run 'cheerful-peacock' - Finished in state Completed()
```

## Resources

For more tips on how to use tasks and flows in a Collection, check out [Using Collections](https://docs.prefect.io/collections/usage/)!

### Installation

Install `prefect-earthdata` with `pip`:

```bash
pip install prefect-earthdata
```

Requires an installation of Python 3.8+.

We recommend using a Python virtual environment manager such as pipenv, conda or virtualenv.

These tasks are designed to work with Prefect 2.0. For more information about how to use Prefect, please refer to the [Prefect documentation](https://docs.prefect.io/).

<!--- ### Saving credentials to block

Note, to use the `load` method on Blocks, you must already have a block document [saved through code](https://docs.prefect.io/concepts/blocks/#saving-blocks) or [saved through the UI](https://docs.prefect.io/ui/blocks/).

Below is a walkthrough on saving block documents through code.

1. Head over to <SERVICE_URL>.
2. Login to your <SERVICE> account.
3. Click "+ Create new secret key".
4. Copy the generated API key.
5. Create a short script, replacing the placeholders (or do so in the UI).

```python
from prefect_earthdata import Block
Block(api_key="API_KEY_PLACEHOLDER").save("BLOCK_NAME_PLACEHOLDER")
```

Congrats! You can now easily load the saved block, which holds your credentials:

```python
from prefect_earthdata import Block
Block.load("BLOCK_NAME_PLACEHOLDER")
```

!!! info "Registering blocks"

    Register blocks in this module to
    [view and edit them](https://docs.prefect.io/ui/blocks/)
    on Prefect Cloud:

    ```bash
    prefect block register -m prefect_earthdata
    ```

A list of available blocks in `prefect-earthdata` and their setup instructions can be found [here](https://giorgiobasile.github.io/prefect-earthdata/blocks_catalog).

--->

### Feedback

If you encounter any bugs while using `prefect-earthdata`, feel free to open an issue in the [prefect-earthdata](https://github.com/giorgiobasile/prefect-earthdata) repository.

If you have any questions or issues while using `prefect-earthdata`, you can find help in either the [Prefect Discourse forum](https://discourse.prefect.io/) or the [Prefect Slack community](https://prefect.io/slack).

Feel free to star or watch [`prefect-earthdata`](https://github.com/giorgiobasile/prefect-earthdata) for updates too!

### Contributing

If you'd like to help contribute to fix an issue or add a feature to `prefect-earthdata`, please [propose changes through a pull request from a fork of the repository](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork).

Here are the steps:

1. [Fork the repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo#forking-a-repository)
2. [Clone the forked repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo#cloning-your-forked-repository)
3. Install the repository and its dependencies:
```
pip install -e ".[dev]"
```
4. Make desired changes
5. Add tests
6. Insert an entry to [CHANGELOG.md](https://github.com/giorgiobasile/prefect-earthdata/blob/main/CHANGELOG.md)
7. Install `pre-commit` to perform quality checks prior to commit:
```
pre-commit install
```
8. `git commit`, `git push`, and create a pull request


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Added

### Changed

### Deprecated

### Removed

### Fixed

- Handle earthaccess returning full file paths on download since >=0.7.0 - [#35](https://github.com/giorgiobasile/prefect-earthdata/pull/35)

### Security

## 0.1.3

Released on October 21st, 2023.

### Fixed

- Fix pydantic imports to handle v2 support from Prefect - [#30](https://github.com/giorgiobasile/prefect-earthdata/pull/30)

## 0.1.2

Released on July 26th, 2023.

### Fixed

- Fixed (again) wrong import of `EarthdataCredentials` in code examples - [#20](https://github.com/giorgiobasile/prefect-earthdata/issues/20)

## 0.1.1

Released on July 25th, 2023.

### Fixed

- Fixed wrong import of `EarthdataCredentials` in code examples - [#20](https://github.com/giorgiobasile/prefect-earthdata/issues/20)

## 0.1.0

Released on July 5th, 2023.

### Added

- Added NASA Earthdata credentials block - [#4](https://github.com/giorgiobasile/prefect-earthdata/issues/4)
- Added Prefect tasks to search and download data through NASA Earthdata - [#8](https://github.com/giorgiobasile/prefect-earthdata/issues/8)


prefect_eo/earthdata.py
"""Module handling NASA Earthdata credentials and tasks"""

import os
from typing import List

import earthaccess
from prefect import get_run_logger, task
from prefect.blocks.core import Block
from pydantic import Field, SecretStr


class EarthdataCredentials(Block):
    """
    Block used to manage authentication with NASA Earthdata.
    NASA Earthdata authentication is handled via the `earthaccess` module.
    Refer to the [earthaccess docs](https://nsidc.github.io/earthaccess/)
    for more info about the possible credential configurations.

    Example:
        Load stored Earthdata credentials:
        ```python
        from prefect_eo import EarthdataCredentials

        ed_credentials_block = EarthdataCredentials.load("BLOCK_NAME")
        ```
    """  # noqa E501

    _logo_url = "https://yt3.googleusercontent.com/ytc/AGIKgqPjIUeAw3_hrkHWZgixdwD5jc-hTWweoCA6bJMhUg=s176-c-k-c0x00ffffff-no-rj"  # noqa
    _block_type_name = "NASA Earthdata Credentials"
    _documentation_url = "https://nsidc.github.io/earthaccess/"  # noqa

    earthdata_username: str = Field(
        default=...,
        description="The Earthdata username of a specific account.",
        title="Earthdata username",
    )
    earthdata_password: SecretStr = Field(
        default=...,
        description="The Earthdata password of a specific account.",
        title="Earthdata password",
    )

    def login(self) -> earthaccess.Auth:
        """
        Returns an authenticated session with NASA Earthdata using
        the [`earthaccess.login()`](https://nsidc.github.io/earthaccess/user-reference/api/api/#earthaccess.api.login) function

        Example:
            Authenticates with NASA Earthdata using the credentials.

            ```python
            from prefect_eo.earthdata import EarthdataCredentials

            earthdata_credentials_block = EarthdataCredentials(
                earthdata_username = "username",
                earthdata_password = "password"
            )
            earthdata_auth = earthdata_credentials_block.login()
            ```
        """  # noqa E501

        os.environ["EARTHDATA_USERNAME"] = self.earthdata_username
        os.environ["EARTHDATA_PASSWORD"] = self.earthdata_password.get_secret_value()
        return earthaccess.login(strategy="environment")


@task
async def earthdata_search_data(
    credentials: EarthdataCredentials, *args, **kwargs
) -> List[earthaccess.results.DataGranule]:
    """
    Searches for data on NASA Earthdata using the
    [`earthaccess.search_data()`](https://nsidc.github.io/earthaccess/user-reference/api/api/#earthaccess.api.search_data) function

    Args:
        credentials: An `EarthdataCredentials` object used
            to authenticate with NASA Earthdata.
        args: Additional positional arguments to be passed
            to `earthaccess.search_data()`.
        kwargs: Additional keyword arguments to be passed
            to `earthaccess.search_data()`.

    Returns:
        A list of `DataGranule` objects representing the search results.

    Example:
        Searches granules through NASA Earthdata.

        ```python
        from prefect import flow
        from prefect_eo.earthdata import EarthdataCredentials
        from prefect_eo.earthdata import earthdata_search_data

        @flow
        def example_earthdata_search_flow():

            credentials = EarthdataCredentials(
                earthdata_userame = "username",
                earthdata_password = "password"
            )

            search_results = earthdata_search_data(
                earthdata_credentials,
                count=1,
                short_name="ATL08",
                bounding_box=(-92.86, 16.26, -91.58, 16.97),
            )
            return search_results

        example_earthdata_search_flow()
        ```
    """  # noqa: E501

    logger = get_run_logger()

    logger.debug("Authenticating to NASA Earthdata")
    auth = credentials.login()
    if not auth.authenticated:
        raise ValueError("Could not authenticate to NASA Earthdata")

    return earthaccess.search_data(*args, **kwargs)


@task
async def earthdata_download(
    credentials: EarthdataCredentials, *args, **kwargs
) -> List[str]:
    """
    Downloads data from NASA Earthdata using the
    [`earthaccess.download()`](https://nsidc.github.io/earthaccess/user-reference/api/api/#earthaccess.api.download) function

    Args:
        credentials: An `EarthdataCredentials` object used
            to authenticate with NASA Earthdata.
        args: Additional positional arguments to be passed
            to `earthaccess.download()`.
        kwargs: Additional keyword arguments to be passed
            to `earthaccess.download()`.

    Returns:
        A list of `DataGranule` objects representing the search results.

    Example:
        Searches and downloads granules through NASA Earthdata.

        ```python
        from prefect import flow
        from prefect_eo.earthdata import EarthdataCredentials
        from prefect_eo.earthdata import earthdata_search_data

        @flow
        def example_earthdata_download_flow():

            credentials = EarthdataCredentials(
                earthdata_userame = "username",
                earthdata_password = "password"
            )

            search_results = earthdata_search_data(
                earthdata_credentials,
                count=1,
                short_name="ATL08",
                bounding_box=(-92.86, 16.26, -91.58, 16.97),
            )

            files = earthdata_download(
                earthdata_credentials,
                granules=granules,
                local_path=download_path
            )

            return granules, files

        example_earthdata_download_flow()
        ```
    """  # noqa: E501

    logger = get_run_logger()

    logger.debug("Authenticating to NASA Earthdata")
    auth = credentials.login()
    if not auth.authenticated:
        raise ValueError("Could not authenticate to NASA Earthdata")

    return earthaccess.download(*args, **kwargs)
